### PR TITLE
Put

### PR DESCRIPTION
**Pull request by Swaagie**
## Summary

Add PUT functionality and use files as attachments. 

## Changelog

Done

## Test Plan

Unit tests are useful as examples on how API's could use this new functionality.

### CODE REVIEW

File: builds.js [hunk 1]
```diff
@@ -113,6 +113,57 @@ class Builds {
...
+   */
+  async put(files, params, fn) {
+    const { env, locale, pkg, version } = this._readParams(params);
+
+    if (!pkg || !env || !version || !files || !Array.isArray(files)) {
+      return fn(new Error('Invalid parameters supplied'));
+    }
+
+    // Read file content and prepare a npm-like body with file attachments.
+    const name = decodeURIComponent(pkg);
```
Comment (epramer-godaddy) [inline builds.js]: why do consumers need to encode the package name when hitting this endpoint?

File: test/builds.test.js [hunk 2]
```diff
@@ -200,6 +200,63 @@ describe('Builds', function () {
     });
   });
 
+  describe('.put', function () {
```
Comment (jcrugzz) [inline test/builds.test.js:203]: @Swaagie there should be a test with a `@scoped/package` to make clear @epramer-godaddy's question. I know we typically need to encode for when it is a URL param so unsure about the decode here

**Review by jcrugzz — CHANGES_REQUESTED** [review]

File: test/builds.test.js (hunk 2 above)
Comment (mswaagman-godaddy) [reply test/builds.test.js:203]: The encode happens in `_readParams` so this is me just inverting that again. Using that function made sense for the other params. I'll add the test though ;)

File: builds.js (hunk 1 above)
Comment (mswaagman-godaddy) [reply builds.js]: No need to, see other comment. I'll add a test. But the `_readParams` by default does `encodeURIComponent`. The alternative is that we use `params.pkg` directly. Will update.

File: builds.js [hunk 3]
```diff
@@ -113,6 +113,57 @@ class Builds {
...
+  async put(files, params, fn) {
+    const { env, locale, pkg, version } = this._readParams(params);
+
+    if (!pkg || !env || !version || !files || !Array.isArray(files)) {
+      return fn(new Error('Invalid parameters supplied'));
+    }
+
+    // Read file content and prepare a npm-like body with file attachments.
+    const name = params.pkg;
+    await this.warehouse.files.getAttachments(files).then(attachments => {
```
Comment (jcrugzz) [inline builds.js:133]: if we are using `.then` to make this callback based to be consistent with other routes still, we shouldnt need this await or this to be an async function.

Comment (jcrugzz) [inline builds.js]: if we are using `.then` to make this callback based to be consistent with other routes still, we shouldnt need this await or this to be an async function.

**Review by jcrugzz — DISMISSED** [review]

**Review by epramer-godaddy — APPROVED** [review]

File: builds.js (hunk 3 above)
Comment (mswaagman-godaddy) [reply builds.js]: O right good point, had a slightly different implementation first, but can obviously just directly use the promise returned from `getAttachments`

**Pull request merged**